src/operator/tensor/dot-inl.h

```diff
@@ -793,16 +764,21 @@ inline void DotCsrRspDnsImpl(const OpContext& ctx,
             mxnet_op::Kernel<mxnet_op::set_zero, cpu>::Launch(s, num_threads,
                                                               ret->dptr<DType>());
           }
-          num_threads = mxnet_op::get_num_threads<cpu>(ret->shape_[0]);
-          dim_t seg_len = (ret->shape_[0] + num_threads - 1) / num_threads;
+          num_threads = ret->shape_[0];
           if (trans_lhs) {
             LOG(FATAL) << "DotCsrRspDnsImpl has not implemented dot(csr.T, rsp) = dns yet";
           } else {
+            const RType* row_idx_ptr = row_idx_r.dptr<RType>();
+            std::unordered_map<RType, dim_t> row_idx_map;
+            row_idx_map.reserve(nnr * 7);
+            for (dim_t ind = 0; ind < nnr; ind++) {
+              row_idx_map.emplace(row_idx_ptr[ind], ind);
+            }
```

Comment:
if nnr is large, constructing the hashtable can take a long time.

Reply:
also, hash tables, while O(1) access time, are pretty expensive for each access (creating the hash) for something like this where you may be doing possibly millions of iterations each pass.  Is there any way to not use a hash table, possibly using more memory?

Reply:
we can use an array to map row id to the physical location. It will take O(N) memory. we can potentially use a tree (e.g., B+tree) to map row ids. In this case, we can trade off the number of memory accesses with memory size.

Reply:
The real number of rows is from 1 million to 8 million in the customer case. Using an array would waste a lot of memory if `nnr` is far less than the real number of rows. Using a tree based data structure is in nature similar to the binary search approach that was originally implemented. I think we should consider using different approaches for different dimension configurations after thorough benchmarks.

Reply:
Agree with @cjolivier01 that O(1) access time doesn't mean instant access. There's always some constant factor like cache misses. However, "possibly millions of iterations" is too exaggerated - as we know the number of elements ahead of time and reserved the number of buckets accordingly, it won't take that many steps.

@zheng-da using an array means you have to set the initial value of the array. We observed that such array is every expensive to maintain in when array size is a few million in _backward_dot. Tree based approach is quite similar to the original implementation. 

We definitely need more benchmark data

Reply:
@eric-haibin-lin if we maintain an array for mapping and an array for recording the number of non-zero entries in the mapping array, the overhead of cleaning the mapping is O(nnr).

@reminisce depending on how we use a tree to store the index. we can achieve O(log_k N) accesses. in practice, k is a large number (e.g., 128), so it's 3 or 4 memory accesses before finding the right location. memory overhead is O(min(N, k * nnr)). it might be better than using a single array or hashtable.

Reply:
btw, the "slowness" I was referring to was not the cache misses, but the hashing function itself done for each access, which is far from free.